### PR TITLE
Imap support (and libpam)

### PR DESCRIPTION
**Pull request by DubbleClick**
This is terrible, it doesn't support ZTS which already disqualifies it for my use case. I'll try to finish this anyway, though.

### CODE REVIEW

File: src/SPC/builder/unix/library/imap.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+{
+    /**
+     * @throws FileSystemException
+     * @throws RuntimeException
+     * @throws WrongUsageException
+     */
+    protected function build(): void
+    {
+        // ext-imap is NOT thread safe, do not build with ZTS builds
+        if ($this->builder->getOption('enable-zts')) {
```
Comment (DubbleClick) [inline src/SPC/builder/unix/library/imap.php:21]: we don't have a "lib-conflicts" option so this is what I came up with

File: src/SPC/builder/unix/library/imap.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                "{$this->builder->configure_env} make {$platform}" .
+                ($this->builder->getLib('openssl') === null ? ' SSLTYPE=none' : '')
+            );
+        shell()
+            ->exec("cp -rf {$this->source_dir}/c-client/c-client.a " . BUILD_LIB_PATH . '/libc-client.a')
+            ->exec("cp -rf {$this->source_dir}/c-client/*.h " . BUILD_INCLUDE_PATH)
+            ->exec("cp -rf {$this->source_dir}/src/osdep/unix/*.h " . BUILD_INCLUDE_PATH);
+    }
+
+    private function replace_file_contents(string|array $filepath, string $old, string $new): void
```
Comment (DubbleClick) [inline src/SPC/builder/unix/library/imap.php:92]: where should I refactor this?

File: src/SPC/builder/unix/library/imap.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        shell()->cd($this->source_dir)
+            ->exec('touch ip6')
+            ->exec(
+                "{$this->builder->configure_env} make {$platform}" .
+                ($this->builder->getLib('openssl') === null ? ' SSLTYPE=none' : '')
+            );
+        shell()
+            ->exec("cp -rf {$this->source_dir}/c-client/c-client.a " . BUILD_LIB_PATH . '/libc-client.a')
+            ->exec("cp -rf {$this->source_dir}/c-client/*.h " . BUILD_INCLUDE_PATH)
+            ->exec("cp -rf {$this->source_dir}/src/osdep/unix/*.h " . BUILD_INCLUDE_PATH);
```
Comment (DubbleClick) [inline src/SPC/builder/unix/library/imap.php:89]: uw-imap doesn't have a 'make install' action defined, so we have to copy files manually

File: src/SPC/builder/unix/library/imap.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            );
+            $this->replace_file_contents(
+                $this->source_dir . '/src/osdep/unix/Makefile',
+                '-lcrypto -lz',
+                '-lcrypto'
+            );
+            $this->replace_file_contents(
+                $this->source_dir . '/src/osdep/unix/Makefile',
+                '-lcrypto',
+                '-lcrypto -lz'
```
Comment (crazywhalecc) [inline src/SPC/builder/unix/library/imap.php:48]: We have `FileSystem::replaceFileStr('/path/to/file', 'src str', 'dst str');`. See https://github.com/crazywhalecc/static-php-cli/blob/main/src/SPC/store/FileSystem.php

File: src/SPC/builder/macos/library/imap.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * http://license.coscl.org.cn/MulanPSL2
+ *
+ * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
+ *
+ * See the Mulan PSL v2 for more details.
+ */
```
[the remark this answers is not in the document]
Comment (DubbleClick) [reply src/SPC/builder/macos/library/imap.php:17]: Oh sorry, it's just a WIP and I copy-pasted code from another extension. Not planning on any copyright related to my contributions here!

File: config/ext.json [hunk 6]
```diff
@@ -132,12 +132,13 @@
...
     "imap": {
         "type": "builtin",
-        "arg-type": "with",
+        "arg-type": "custom",
         "lib-depends": [
             "imap"
         ],
         "lib-suggests": [
-            "kerberos"
+            "kerberos",
```
[the remark this answers is not in the document]
Comment (DubbleClick) [reply config/ext.json:140]: It doesn't rely on it I believe, but it's still suggested to use it. I will come back to it when I finish the PR and remove the WIP from the title.

File: src/SPC/builder/unix/library/imap.php (hunk 4 above)
Comment (DubbleClick) [reply src/SPC/builder/unix/library/imap.php:48]: Thank you, will refactor to use that

**Comment by DubbleClick** [comment]
```
configure:44947: checking whether build with IMAP works
configure:44987: musl-gcc -o conftest  -fvisibility=hidden  -D_GNU_SOURCE -I/opt/static-php-cli/buildroot/include  -Wl,-rpath,/opt/static-php-cli/buildroot/lib -L/opt/static-php-cli/buildroot/lib conftest.c -lc-client   -lcrypt -lrt -lm -ldl -lpthread -lssl -lcrypto -lz -lssl -lcrypto >&5
/usr/bin/ld: /opt/static-php-cli/buildroot/lib/libc-client.a(osdep.o): in function `checkpw_cleanup':
/opt/static-php-cli/source/uwimap/c-client/osdep.c:144: undefined reference to `pam_setcred'
/usr/bin/ld: /opt/static-php-cli/buildroot/lib/libc-client.a(osdep.o): in function `checkpw':
/opt/static-php-cli/source/uwimap/c-client/osdep.c:164: undefined reference to `pam_start'
/usr/bin/ld: /opt/static-php-cli/source/uwimap/c-client/osdep.c:166: undefined reference to `pam_set_item'
/usr/bin/ld: /opt/static-php-cli/source/uwimap/c-client/osdep.c:167: undefined reference to `pam_authenticate'
/usr/bin/ld: /opt/static-php-cli/source/uwimap/c-client/osdep.c:168: undefined reference to `pam_acct_mgmt'
/usr/bin/ld: /opt/static-php-cli/source/uwimap/c-client/osdep.c:169: undefined reference to `pam_setcred'
/usr/bin/ld: /opt/static-php-cli/buildroot/lib/libc-client.a(osdep.o): in function `checkpw_cleanup':
/opt/static-php-cli/source/uwimap/c-client/osdep.c:144: undefined reference to `pam_setcred'
/usr/bin/ld: /opt/static-php-cli/source/uwimap/c-client/osdep.c:145: undefined reference to `pam_end'
/usr/bin/ld: /opt/static-php-cli/source/uwimap/c-client/osdep.c:145: undefined reference to `pam_end'
collect2: error: ld returned 1 exit status
```

php-src/configure adds -lpam for a *shared* build of ext-imap, but not the a static build of imap. Yet, the static build of imap apparently still relies on libpam. I'm not sure if it's possible to compile this extension statically.

File: src/SPC/builder/unix/library/imap.php (hunk 1 above)
Comment (DubbleClick) [reply src/SPC/builder/unix/library/imap.php]: I'm not sure if this is actually true. The php docs say it's not thread safe, but there are references in the extension source. I've uncommented it for now.

**Comment by DubbleClick** [comment]
I think I will drop this here. Too much of a headache to support an extension that doesn't have clear documentation, relies on a long abandoned c-client and doesn't even compile statically. I refactored to using webklex\php-imap in my projects instead.

If someone would like to pick this up, please do.

**Comment by DubbleClick** [comment]
I forgot to mention, there's a fork of panda-imap (fork of uw-imap that Marc Crispin maintained until he passed away) that is still maintained.

https://github.com/nkhorman/panda-imap

Same problem when compiling php-imap staticallywith libpam references, though.

**Comment by crazywhalecc** [comment]
> I think I will drop this here. Too much of a headache to support an extension that doesn't have clear documentation, relies on a long abandoned c-client and doesn't even compile statically. I refactored to using webklex\php-imap in my projects instead.

😢 Me too. I have tried what you mentioned before, but in the end it was difficult to even compile Kerberos, so I don’t have time to deal with them yet.

**Comment by crazywhalecc** [comment]
I don't know if it compiles out the `.a` library on macOS and links correctly. Because the location of the built-in library in the macOS system is stable, you can consider using some dynamic links to support it on mac. But if you have already made a series of difficult but ineffective efforts on the macOS platform, I think we can give up.

**Comment by DubbleClick** [comment]
Since MacOS doesn't allow entirely static binaries and can load .so's during compilation, we could patch the php-src ./configure file to link it with `-lpam -lm -l...` dependencies.

It might work. The same would probably also work on Linux, but what's the point? It wouldn't be *static*-php anymore. The real fix would be to get libc-client to compile in all the pam_... functions, but there are no static libraries for pam...

**Comment by DubbleClick** [comment]
I'd try again if I had a good idea on how to go forward. If I were to get linux-pam compiled statically, I think php still wouldn't build with imap because autoconf creates a ./Configure that doesn't add -lpam. Or would it be enough if I built uw-imap again with -lpam, aka would that statically compile the pam_... functions into libc-client? It doesn't do it for openssl & zlib...

**Comment by crazywhalecc** [comment]
> I see very little point in macOS only extensions because practically no server serving php will be a mac. Web is still php's primary purpose.

Sometimes we need to use these extensions on mac because they can be used as a local, quick-installation development environment (Laravel Herd is based on this).

> I'd try again if I had a good idea on how to go forward. If I were to get linux-pam compiled statically, I think php still wouldn't build with imap because autoconf creates a ./Configure that doesn't add -lpam. Or would it be enough if I built uw-imap again with -lpam, aka would that statically compile the pam_... functions into libc-client? It doesn't do it for openssl & zlib...

The current build process works well for dependent libraries that automatically use pkg-config during the autoconf or cmake stages. For libraries that only support dynamic compilation by default, we may not be able to avoid using the process of replacing `-lxxx` to `libxxx.a`. And for static compilation, if the imap library depends on libpam, its `.a` library only has its symbols, which will only be searched and linked when PHP is finally built.

**Comment by DubbleClick** [comment]
> we may not be able to avoid using the process of replacing -lxxx to libxxx.a

Is it important to specify them that way? I just use -lssl -lcrypto -lz in openldap and it works fine - or does patchPkgConfig do this? -lz is listed as a private dependency in `libldap.pc`.

I created a `libimap.pc` manually with the respective contents amd called patchPkgConfif, but it didn't seem to have an effect on php-src compilation (-lpam was still not added in php-src ./configure).

**Comment by crazywhalecc** [comment]
I'm not very clear about the specific working principle of the linker, but after many experiments I found that it still depends on the dependency library itself's support for static compilation and whether the config.m4 of the PHP extension is well written.

I sometimes encounter link errors. There are several possibilities:

1. The `xxx.pc` of the library is not patched (that is, the `prefix=` field of the pkgconf file)
2. The library itself does not support static compilation well, for example, it does not support `--enable-static`, etc.
3. The library itself has no clear dependencies (currently, the `.a` library dependency for statically compiled PHP depends on the `lib-depends` and `static-libs` fields of the configuration file, which are very important)

**Comment by DubbleClick** [comment]
> The current compilation process will theoretically recursively obtain all dependent xxx.a files. When linking PHP at the end, if libxxx.a can be seen in the command line in debug mode but there is still a link error, then There may be a real problem.

In the case of ldap, it still complained about unresolved external reference to zlib functions even though libz.a was indeed in ${prefix}/lib. Adding the zlib extension fixed it and finally convinced php to build properly.

> TBH I really should blog about all the problems I encounter when refactoring this project, it might be helpful in the future. Also, I don't want you to get stuck solving recursive compilation problems

That's a great idea. It's honestly unfathomable that the official documentation for building php with extensions (especially static) is this poor.

**Comment by DubbleClick** [comment]
Okay, I got linux-pam to compile, but only with gcc. Musl fails because

```sh
fatal error: linux/quota.h: No such file or directory
   13 | #include <linux/quota.h>
```
which is probably because musl-gcc doesn't ship with kernel headers. You probably know the answer how to fix this? (on alpine it might just be apt install linux-headers). @crazywhalecc

**Comment by DubbleClick** [comment]
PAM compiles, PHP (after patching ./configure) configures, but during build it still complains about undefined references...

```
/bin/sh /opt/static-php-cli/source/php-src/libtool --silent --preserve-dup-deps --tag CC --mode=link musl-gcc -export-dynamic -fno-common -Wformat-truncation -Wlogical-op -Wduplicated-cond -Wno-clobbered -Wall -Wextra -Wno-strict-aliasing -Wno-unused-parameter -Wno-sign-compare -fvisibility=hidden -Wimplicit-fallthrough=1 -DZEND_SIGNALS  -g -Os -fno-ident -Xcompiler -march=corei7 -Xcompiler -mtune=core-avx2 -all-static  -R /opt/static-php-cli/buildroot/lib ext/date/php_date.o ext/date/lib/astro.o ext/date/lib/dow.o ext/date/lib/parse_date.o ext/date/lib/parse_tz.o ext/date/lib/parse_posix.o ext/date/lib/timelib.o ext/date/lib/tm2unixtime.o ext/date/lib/unixtime2tm.o ext/date/lib/parse_iso_intervals.o ext/date/lib/interval.o ext/pcre/pcre2lib/pcre2_auto_possess.o ext/pcre/pcre2lib/pcre2_chartables.o ext/pcre/pcre2lib/pcre2_compile.o ext/pcre/pcre2lib/pcre2_config.o ext/pcre/pcre2lib/pcre2_context.o ext/pcre/pcre2lib/pcre2_dfa_match.o ext/pcre/pcre2lib/pcre2_error.o ext/pcre/pcre2lib/pcre2_jit_compile.o ext/pcre/pcre2lib/pcre2_maketables.o ext/pcre/pcre2lib/pcre2_match.o ext/pcre/pcre2lib/pcre2_match_data.o ext/pcre/pcre2lib/pcre2_newline.o ext/pcre/pcre2lib/pcre2_ord2utf.o ext/pcre/pcre2lib/pcre2_pattern_info.o ext/pcre/pcre2lib/pcre2_serialize.o ext/pcre/pcre2lib/pcre2_string_utils.o ext/pcre/pcre2lib/pcre2_study.o ext/pcre/pcre2lib/pcre2_substitute.o ext/pcre/pcre2lib/pcre2_substring.o ext/pcre/pcre2lib/pcre2_tables.o ext/pcre/pcre2lib/pcre2_ucd.o ext/pcre/pcre2lib/pcre2_valid_utf.o ext/pcre/pcre2lib/pcre2_xclass.o ext/pcre/pcre2lib/pcre2_find_bracket.o ext/pcre/pcre2lib/pcre2_convert.o ext/pcre/pcre2lib/pcre2_extuni.o ext/pcre/pcre2lib/pcre2_script_run.o ext/pcre/php_pcre.o ext/hash/hash.o ext/hash/hash_md.o ext/hash/hash_sha.o ext/hash/hash_ripemd.o ext/hash/hash_haval.o ext/hash/hash_tiger.o ext/hash/hash_gost.o ext/hash/hash_snefru.o ext/hash/hash_whirlpool.o ext/hash/hash_adler32.o ext/hash/hash_crc32.o ext/hash/hash_fnv.o ext/hash/hash_joaat.o ext/hash/sha3/generic64lc/KeccakP-1600-opt64.o ext/hash/sha3/generic64lc/KeccakHash.o ext/hash/sha3/generic64lc/KeccakSponge.o ext/hash/hash_sha3.o ext/hash/murmur/PMurHash.o ext/hash/murmur/PMurHash128.o ext/hash/hash_murmur.o ext/hash/hash_xxhash.o ext/imap/php_imap.o ext/json/json.o ext/json/json_encoder.o ext/json/json_parser.o ext/json/json_scanner.o ext/reflection/php_reflection.o ext/spl/php_spl.o ext/spl/spl_functions.o ext/spl/spl_iterators.o ext/spl/spl_array.o ext/spl/spl_directory.o ext/spl/spl_exceptions.o ext/spl/spl_observer.o ext/spl/spl_dllist.o ext/spl/spl_heap.o ext/spl/spl_fixedarray.o ext/standard/crypt_freesec.o ext/standard/crypt_blowfish.o ext/standard/crypt_sha512.o ext/standard/crypt_sha256.o ext/standard/php_crypt_r.o ext/standard/array.o ext/standard/base64.o ext/standard/basic_functions.o ext/standard/browscap.o ext/standard/crc32.o ext/standard/crypt.o ext/standard/datetime.o ext/standard/dir.o ext/standard/dl.o ext/standard/dns.o ext/standard/exec.o ext/standard/file.o ext/standard/filestat.o ext/standard/flock_compat.o ext/standard/formatted_print.o ext/standard/fsock.o ext/standard/head.o ext/standard/html.o ext/standard/image.o ext/standard/info.o ext/standard/iptc.o ext/standard/lcg.o ext/standard/link.o ext/standard/mail.o ext/standard/math.o ext/standard/md5.o ext/standard/metaphone.o ext/standard/microtime.o ext/standard/pack.o ext/standard/pageinfo.o ext/standard/quot_print.o ext/standard/rand.o ext/standard/mt_rand.o ext/standard/soundex.o ext/standard/string.o ext/standard/scanf.o ext/standard/syslog.o ext/standard/type.o ext/standard/uniqid.o ext/standard/url.o ext/standard/var.o ext/standard/versioning.o ext/standard/assert.o ext/standard/strnatcmp.o ext/standard/levenshtein.o ext/standard/incomplete_class.o ext/standard/url_scanner_ex.o ext/standard/ftp_fopen_wrapper.o ext/standard/http_fopen_wrapper.o ext/standard/php_fopen_wrapper.o ext/standard/credits.o ext/standard/css.o ext/standard/var_unserializer.o ext/standard/ftok.o ext/standard/sha1.o ext/standard/user_filters.o ext/standard/uuencode.o ext/standard/filters.o ext/standar/usr/bin/ld: /opt/static-php-cli/buildroot/lib/libc-client.a(osdep.o): in function `checkpw_cleanup':
/opt/static-php-cli/source/imap/c-client/osdep.c:144: undefined reference to `pam_setcred'
/usr/bin/ld: /opt/static-php-cli/buildroot/lib/libc-client.a(osdep.o): in function `checkpw':
/opt/static-php-cli/source/imap/c-client/osdep.c:164: undefined reference to `pam_start'
/usr/bin/ld: /opt/static-php-cli/source/imap/c-client/osdep.c:166: undefined reference to `pam_set_item'
/usr/bin/ld: /opt/static-php-cli/source/imap/c-client/osdep.c:167: undefined reference to `pam_authenticate'
/usr/bin/ld: /opt/static-php-cli/source/imap/c-client/osdep.c:168: undefined reference to `pam_acct_mgmt'
/usr/bin/ld: /opt/static-php-cli/source/imap/c-client/osdep.c:169: undefined reference to `pam_setcred'
/usr/bin/ld: /opt/static-php-cli/buildroot/lib/libc-client.a(osdep.o): in function `checkpw_cleanup':
/opt/static-php-cli/source/imap/c-client/osdep.c:144: undefined reference to `pam_setcred'
/usr/bin/ld: /opt/static-php-cli/source/imap/c-client/osdep.c:145: undefined reference to `pam_end'
/usr/bin/ld: /opt/static-php-cli/source/imap/c-client/osdep.c:145: undefined reference to `pam_end'
collect2: error: ld returned 1 exit status
make: *** [Makefile:258: sapi/cli/php] Error 1
```
Which is weird, given that 

```
[root@localhost static-php-cli]# nm /opt/static-php-cli/buildroot/lib/libpam.a | grep -A 10 -B 6 pam_setcred
pam_auth.o:
                 U _pam_await_timer
                 U _pam_dispatch
                 U _pam_sanitize
                 U _pam_start_timer
0000000000000000 T pam_authenticate
00000000000000d0 T pam_setcred
                 U syslog

pam_data.o:
00000000000001b0 T _pam_free_data
                 U _pam_strdup
                 U free
                 U malloc
0000000000000120 T pam_get_data
0000000000000000 T pam_set_data
                 U pam_syslog
```

I'm not sure what the issue is here, maybe you have an idea?

**Comment by DubbleClick** [comment]
Using ar to combine libc-client.a and libpam.a into a new libc-client.a makes php compile...

```sh
mv libc-client.a oldlibc-client.a
ar -M <<EOM
    CREATE libc-client.a
    ADDLIB liboldc-client.a
    ADDLIB libpam.a
    SAVE
    END
EOM
ranlib libc-client.a
```

```sh
[18:05:50] [INFO] running cli sanity check
[18:05:50] [INFO] [EXEC] /opt/static-php-cli/buildroot/bin/php -r "echo \"hello\";"
[18:05:50] [DEBU] testing ext: imap
[18:05:50] [DEBU] Running command with result: /opt/static-php-cli/buildroot/bin/php --ri "imap"
[18:05:50] [INFO] Build complete, used 121.581 s !
[18:05:50] [INFO] Static php binary path: /opt/static-php-cli/buildroot/bin/php
[18:05:50] [DEBU] Making new directory recursive: /opt/static-php-cli/buildroot/license
[18:05:50] [INFO] License path: /opt/static-php-cli/buildroot/license/
[root@localhost static-php-cli]# ./buildroot/bin/php -m
[PHP Modules]
Core
date
hash
imap
json
pcre
Reflection
SPL
standard
```

**Comment by DubbleClick** [comment]
Turns out it's an issue in c-client that uses gss/pam without correctly linking to it. `nm libc-client.a | grep pam_` shows references to pam_ functions without an address.
Building uw-imap for slx works.

Time to tidy up the mess and we should be good to go.

**Comment by DubbleClick** [comment]
There are a few comments in the code that you're better qualified to handle. Shouldn't be too much work. Keep in mind that if you compile imap and openssl, you also have to enable zlib (you added it to the require of openssl on the ldap branch, but it's not here yet)

File: src/SPC/builder/unix/library/imap.php [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    /**
+     * @throws FileSystemException
+     * @throws RuntimeException
+     * @throws WrongUsageException
+     */
+    protected function build(): void
+    {
+        //        ext-imap is NOT thread safe, do not build with ZTS builds
+        //        if ($this->builder->getOption('enable-zts')) {
+        //            return;
```
Comment (stloyd) [inline src/SPC/builder/unix/library/imap.php:24]: I would add this `if` but throw a runtime exception with the error like in the comment instead.

File: src/SPC/builder/unix/library/imap.php [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+declare(strict_types=1);
+
+namespace SPC\builder\unix\library;
+
+use SPC\exception\FileSystemException;
+use SPC\exception\RuntimeException;
+use SPC\exception\WrongUsageException;
+use SPC\store\FileSystem;
+use function Safe\xdiff_file_patch_binary;
```
Comment (stloyd) [inline src/SPC/builder/unix/library/imap.php:11]: I don't see any usage of this.

Comment (DubbleClick) [reply src/SPC/builder/unix/library/imap.php]: woopsie, I considered replacing the xdiff_file_patch method from the xdiff extension, but felt like introducing that requirement was not my decision to make

File: src/SPC/builder/unix/library/imap.php (hunk 7 above)
Comment (DubbleClick) [reply src/SPC/builder/unix/library/imap.php]: made it a wrong usage extension, as the user should make sure to pass compatible options

File: src/SPC/builder/unix/library/imap.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                <<<'EOL'
+if (!(stream->con = (SSL *) SSL_new (stream->context)))
+    return "SSL connection failed";
+#if OPENSSL_VERSION_NUMBER >= 0x10101000
+  /* Use SNI in case server requires it with TLSv1.3.
+   * Literal IP addresses not permitted per RFC 6066. */
+  if (!a2i_IPADDRESS(host)) {
+    ERR_clear_error();
+    SSL_set_tlsext_host_name(stream->con,host);
+  }
```
Comment (crazywhalecc) [inline src/SPC/builder/unix/library/imap.php:63]: Actually I prefer using `patch` command to patch the multi-line content if it is not dynamic. Related issue: #206. I will write a util class to do it. We can make a `xxx.patch`.

**Comment by crazywhalecc** [comment]
> There are a few comments in the code that you're better qualified to handle. Shouldn't be too much work. Keep in mind that if you compile imap and openssl, you also have to enable zlib (you added it to the require of openssl on the ldap branch, but it's not here yet)

I will merge it after running on Linux asap. Documentation bothering me..

**Comment by DubbleClick** [comment]
Don't worry about the docs, I'll do those later today. Maybe I'll even add ext-pam on Monday, will have to see how much time I have.

File: src/SPC/builder/unix/library/imap.php (hunk 9 above)
Comment (DubbleClick) [reply src/SPC/builder/unix/library/imap.php:63]: There is a xdiff php extension that includes patching functionality. Might not have to create a class for this.

Comment (crazywhalecc) [reply src/SPC/builder/unix/library/imap.php:63]: We can default to the patch command being installed on the system (the doctor command check also has this item), and there is no need to introduce additional dependencies to spc.

**Comment by crazywhalecc** [comment]
I cannot compile imap sucessfully on my Debian VM:

```
osdep.c:482:49: error: invalid use of incomplete typedef ‘X509’ {aka ‘struct x509_st’}
  482 |     if (scq) return (*scq) (err,host,cert ? cert->name : "???") ? NIL : "";
      |                                                 ^~
osdep.c:484:51: error: invalid use of incomplete typedef ‘X509’ {aka ‘struct x509_st’}
  484 |     sprintf (tmp,"*%.128s: %.255s",err,cert ? cert->name : "???");
      |                                                   ^~
osdep.c: In function ‘ssl_validate_cert’:
osdep.c:534:17: error: invalid use of incomplete typedef ‘X509’ {aka ‘struct x509_st’}
  534 |   else if (!cert->name) ret = "No name in certificate";
      |                 ^~
osdep.c:536:28: error: invalid use of incomplete typedef ‘X509’ {aka ‘struct x509_st’}
  536 |   else if (s = strstr (cert->name,"/CN=")) {
      |                            ^~
osdep.c: In function ‘ssl_server_init’:
osdep.c:915:12: warning: ‘TLSv1_server_method’ is deprecated: Since OpenSSL 1.1.0 [-Wdeprecated-declarations]
  915 |            TLSv1_server_method () :
      |            ^~~~~~~~~~~~~~~~~~~
In file included from osdep.c:231:
/home/jerry/static-php-cli/buildroot/include/openssl/ssl.h:2019:50: note: declared here
 2019 | OSSL_DEPRECATEDIN_1_1_0 __owur const SSL_METHOD *TLSv1_server_method(void);
      |                                                  ^~~~~~~~~~~~~~~~~~~
osdep.c:930:5: warning: ‘SSL_CTX_use_RSAPrivateKey_file’ is deprecated: Since OpenSSL 3.0 [-Wdeprecated-declarations]
  930 |     else if (!(SSL_CTX_use_RSAPrivateKey_file (stream->context,key,
      |     ^~~~
In file included from osdep.c:231:
/home/jerry/static-php-cli/buildroot/include/openssl/ssl.h:1700:12: note: declared here
 1700 | __owur int SSL_CTX_use_RSAPrivateKey_file(SSL_CTX *ctx, const char *file,
      |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
osdep.c: In function ‘ssl_genkey’:
osdep.c:986:5: warning: ‘RSA_generate_key’ is deprecated: Since OpenSSL 0.9.8 [-Wdeprecated-declarations]
  986 |     if (!(key = RSA_generate_key (export ? keylength : 1024,RSA_F4,NIL,NIL))) {
      |     ^~
In file included from /home/jerry/static-php-cli/buildroot/include/openssl/x509.h:36,
                 from /home/jerry/static-php-cli/buildroot/include/openssl/ssl.h:31,
                 from osdep.c:231:
/home/jerry/static-php-cli/buildroot/include/openssl/rsa.h:259:30: note: declared here
  259 | OSSL_DEPRECATEDIN_0_9_8 RSA *RSA_generate_key(int bits, unsigned long e, void
      |                              ^~~~~~~~~~~~~~~~
make[3]: *** [Makefile:926：osdep.o] 错误 1
make[3]: 离开目录“/home/jerry/static-php-cli/source/imap/c-client”
make[2]: *** [Makefile:677：slx] 错误 2
make[2]: 离开目录“/home/jerry/static-php-cli/source/imap/c-client”
make[1]: *** [Makefile:688：OSTYPE] 错误 2
make[1]: 离开目录“/home/jerry/static-php-cli/source/imap”
make: *** [Makefile:311：slx] 错误 2
```

**Comment by DubbleClick** [comment]
Did the patches run through successfully?

**Comment by crazywhalecc** [comment]
Yeah. I tried your old patch code and got same error.

image

**Comment by DubbleClick** [comment]
That's a beautiful colour scheme. I'll install debian and see if I can solve the problems.

**Comment by crazywhalecc** [comment]
> Maybe it can't build for "slx".

I noticed it. Because I just started trying to build it on macOS and got an error, and then I looked at the Makefile, basically it needs to correspond to the current operating system exactly.

**Comment by DubbleClick** [comment]
I think slx should work for every (linux) os since we include -lcyrpto.

**Comment by DubbleClick** [comment]
I'm still stuck on trying to compile pkg-config on debian...

**Comment by DubbleClick** [comment]
I just set up a new Debian 12 (with GUI) VM. I've installed every required package by hand (some of them were missing after running doctor).

image

**Comment by DubbleClick** [comment]
GCC, musl-gcc wrapper and self compiled x86_64-linux-musl all fail to compile glib.

**Comment by crazywhalecc** [comment]
This is strange, all my local Linux testing was done on Debian12. I will try with clean OS.

**Comment by DubbleClick** [comment]
I just did a clean installation because I was tired of it and lo and behold, the same problem.

**Comment by DubbleClick** [comment]
alright, works with and without ssl now

**Comment by DubbleClick** [comment]
Many of the issues with unreferenced symbols when using the embed api are caused by wrong lib include order, even when you pass them to php-src/configure in correct order. I added zlib as a requirement in openssl.pc after libcrypto and now we don't need to force the zlib extension anymore when openssl is enabled.

**Comment by crazywhalecc** [comment]
I got error when building libpam.

```
[01:51:02] [INFO] Building required library [libpam]
[01:51:02] [INFO] Entering dir: /home/jerry/static-php-cli/source/libpam
[01:51:02] [INFO] [EXEC] ./autogen.sh
[01:51:02] [DEBU] Running command with direct output: cd '/home/jerry/static-php-cli/source/libpam' && ./autogen.sh
+ umask 022
+ touch ChangeLog
+ autoreconf -fiv -Wall
autoreconf: Entering directory `.'
autoreconf: running: autopoint --force
Can't exec "autopoint": no such file or directory at /usr/share/autoconf/Autom4te/FileUtils.pm line 345.
autoreconf: failed to run autopoint: No such file or directory
autoreconf: autopoint is needed because this package uses Gettext
```

**Comment by crazywhalecc** [comment]
What a painful debugging process. 😞 It got worse. Not only did I encounter the previous X509 error, but I also encountered the header file `security/pam_*.h` that could not be found, plus the `autopoint` problem above.

I used `apt install autopoint` on Debian and it worked, on Alpine using `apk add gettext` it didn't seem to work.

**Comment by crazywhalecc** [comment]
I'll turn this PR into a draft first. Maybe there's still a lot of work to be done.

**Comment by DubbleClick** [comment]
Unfortunately to get this working on all systems it has to happen in conjunction with the c++ branch I think.

Alpine may need to install gettext-autopoint like rhel.

I will close this branch and get the c++ branch working for Alpine and aarch64 (rhel, debian, alpine). It will be a very large PR though. :/

**Pull request closed without merging**